src/main/java/org/spongepowered/api/resource/ResourcePath.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.util.ResettableBuilder;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * <p>A resource path should contain a namespace. If one is not provided,
+ * {@code minecraft} will be used instead. The namespace and path must not
+ * contain any special characters or uppercase letters.
+ *
+ * <p>In the pack, the path will point to a resource. The resource should
+ * be located roughly at {@code data/namespace/path}</p>
+ *
+ * <p>A resource path should be usable in a {@link Map}, so implementations
+ * should override {@link #hashCode()} and {@link #equals(Object)}.</p>
+ */
+public interface ResourcePath extends Comparable<ResourcePath>, Iterable<ResourcePath> {
+
+    /**
+     * Creates a new {@link Builder} to build a {@link ResourcePath}.
+     *
+     * @return The new builder
+     */
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * Creates a new {@link ResourcePath} using the given namespace and path.
+     *
+     * @param namespace The namespace to use
+     * @param path The path to use
+     * @return A new ResourcePath
+     */
+    static ResourcePath of(String namespace, String path) throws IllegalArgumentException {
```

Comment:
How about other `minecraft(String)` or `plugin(PluginContainer/Object, String)` methods?

Reply:
I don't feel those methods are necessary. You can just as easily stick in `"minecraft`" or `"foobar"`. I prefer it to be specific for this.